File: merge-sort.c
#include <stdio.h>

void merge(int *nums, size_t start, size_t end);
void mergeSort(int *nums, size_t start, size_t end);

int main(void){
    int nums[] = {27, 4, 7, 12, 9};
    size_t numsSize = sizeof(nums)/sizeof(int);
    mergeSort(nums, 0, numsSize -1);
    for(size_t i = 0; i < numsSize; i++)
        printf("%d\t", nums[i]);
    printf("\n");
    return 0;
}

void merge(int *nums, size_t start, size_t end){
    size_t mid = (start + end) / 2;
    int tempArr[end-start+1];
    int i = start, j = mid + 1, k = 0;

    while(i <= mid && j <= end){
        if(nums[i] < nums[j])
            tempArr[k++] = nums[i++];
        else 
            tempArr[k++] = nums[j++];
    }

    while(i <= mid)
        tempArr[k++] = nums[i++];

    while(j <= end)
        tempArr[k++] = nums[j++];

    for(size_t i = 0; i < end + 1; i++)
        nums[start+i] = tempArr[i];
}

void mergeSort(int *nums, size_t start, size_t end){
    if(start == end)
        return;
    size_t mid = (start + end) / 2;
    mergeSort(nums, start, mid);
    mergeSort(nums, mid + 1, end);
    merge(nums, start, end);
}

File: go/algorithms/merge-sort.go
package algorithms

func MergeSort(s []int) []int {
	if len(s) == 1 {
		return s
	}
	return merge(MergeSort(s[:len(s)/2]), MergeSort(s[len(s)/2:]))
}

func merge(s1, s2 []int) []int {
	var i, j int
	s3 := make([]int, 0, len(s1)+len(s2))
	for i < len(s1) && j < len(s2) {
		if s1[i] < s2[j] {
			s3 = append(s3, s1[i])
			i++
		} else {
			s3 = append(s3, s2[j])
			j++
		}
	}
	for i < len(s1) {
		s3 = append(s3, s1[i])
		i++
	}

	for j < len(s2) {
		s3 = append(s3, s2[j])
		j++
	}
	return s3
}
